### PR TITLE
Bugfix/itdev 36629 prepaid expiration calculation

### PR DESCRIPTION
**Pull request by cdeanj2**
Previously an allocation with a prepaid billing start date like "03/31/2025" would result in a calculated prepaid expiration date that does not exist (like "04/31/2025"). Currently users are only allowed to have prepaid start dates on the first of the month. This issue is not currently impacting users, but it has caused unit tests to occasionally fail. These revisions will prevent future failures.

To manually test create a prepaid allocation with any prepaid start date as long as it is on the first of the month and is in the past (like 04/01/2025). Once the allocation is active, run `coldfront run_check_billing_cycles` in the coldfront-dev-1 container (if testing locally) to verify the prepaid expiration date is calculated as expected.

### CODE REVIEW

File: coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py [hunk 1]
```diff
@@ -239,3 +237,27 @@ def test_billing_cycle_manager_monthly(self):
...
+        check_allocation_billing_cycle_and_prepaid_exp()
+
+        date_string = AllocationAttribute.objects.get(
+            allocation=prepaid_allocation,
+            allocation_attribute_type__name="prepaid_expiration",
+        ).value
+        date_format = "%Y-%m-%d"
+
+        print(date_string)
+
```
Comment (melonistic) [inline coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py:258]: @cdeanj2 What is this test trying to verify or assert?

File: coldfront/plugins/qumulo/management/commands/check_billing_cycles.py [hunk 2]
```diff
@@ -61,15 +62,16 @@ def calculate_prepaid_expiration(
...
         prepaid_months = int(prepaid_months)
-        prepaid_until = datetime(
-            prepaid_billing_start.year
-            + (prepaid_billing_start.month + prepaid_months - 1) // 12,
-            (prepaid_billing_start.month + prepaid_months - 1) % 12 + 1,
-            prepaid_billing_start.day,
-        )
+        prepaid_until = prepaid_billing_start + relativedelta(months=prepaid_months)
+
+        if prepaid_billing_start.day == calendar.monthrange(prepaid_until.year, prepaid_until.month)[1]:
```
Comment (peaches7083) [inline coldfront/plugins/qumulo/management/commands/check_billing_cycles.py:71]: As discussed in Slack, this condition doesn't work - if the prepaid period goes from February 28th to June 28th, say, then the condition won't be true, so the code to shift June 28th to June 30th won't run.

File: coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py [hunk 3]
```diff
@@ -98,12 +98,10 @@ def expected_prepaid_expiration_calculation(allocation):
         prepaid_billing_start = datetime.strptime(prepaid_billing_start, "%Y-%m-%d")
         prepaid_months = int(prepaid_months)
 
-        prepaid_until = datetime(
-            prepaid_billing_start.year
-            + (prepaid_billing_start.month + prepaid_months - 1) // 12,
-            (prepaid_billing_start.month + prepaid_months - 1) % 12 + 1,
-            prepaid_billing_start.day,
-        )
+        prepaid_until = prepaid_billing_start + relativedelta(months=prepaid_months)
```
Comment (peaches7083) [inline coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py:101]: This is the logic that should be used above in `check_billing_cycles` I believe.

File: coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py [hunk 4]
```diff
@@ -239,3 +237,27 @@ def test_billing_cycle_manager_monthly(self):
         ).value
```
Comment (peaches7083) [inline coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py:238]: Testing a leap year date (a prepaid expiration that starts on February 29th and another that ends on February 29th) to make sure the date handling is still good would be very useful.

File: coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py (hunk 1 above)
Comment (cdeanj2) [reply coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py:258]: Thanks for catching that, I forgot to take it out

File: coldfront/plugins/qumulo/management/commands/check_billing_cycles.py [hunk 5]
```diff
@@ -61,15 +62,18 @@ def calculate_prepaid_expiration(
         name="prepaid_expiration"
     )
     logger.warn(f"Calculation prepaid expiration")
+    
     if bill_cycle == "prepaid" and prepaid_expiration == None:
         prepaid_billing_start = datetime.strptime(prepaid_billing_start, "%Y-%m-%d")
```
Comment (melonistic) [inline coldfront/plugins/qumulo/management/commands/check_billing_cycles.py:67]: @cdeanj2 what is the datatype of `prepaid_billing_start `?

Comment (melonistic) [reply coldfront/plugins/qumulo/management/commands/check_billing_cycles.py:67]: Never mind; it's a string.

File: coldfront/plugins/qumulo/management/commands/check_billing_cycles.py [hunk 6]
```diff
@@ -61,15 +62,18 @@ def calculate_prepaid_expiration(
...
+            last_day_of_month = True
         prepaid_months = int(prepaid_months)
-        prepaid_until = datetime(
-            prepaid_billing_start.year
-            + (prepaid_billing_start.month + prepaid_months - 1) // 12,
-            (prepaid_billing_start.month + prepaid_months - 1) % 12 + 1,
-            prepaid_billing_start.day,
-        )
+        prepaid_until = prepaid_billing_start + relativedelta(months=prepaid_months)
+        if last_day_of_month == True:
```
Comment (melonistic) [inline coldfront/plugins/qumulo/management/commands/check_billing_cycles.py:73]: @ cdeanj2 Can someone's prepaid plan expire on any day (prepaid_until ) of the month? I thought all plans expire at the end of the month.

Comment (cdeanj2) [reply coldfront/plugins/qumulo/management/commands/check_billing_cycles.py:73]: Currently users are only allowed to select a prepaid start date on the first of the month. This code is here to handle any edge cases that may occur if we decide to allow any prepaid start date in the future.

File: coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py [hunk 7]
```diff
@@ -254,10 +254,30 @@ def test_prepaid_start_31_of_month(self):
...
+        check_allocation_billing_cycle_and_prepaid_exp()
+
+        date_string = AllocationAttribute.objects.get(
+            allocation=prepaid_allocation,
+            allocation_attribute_type__name="prepaid_expiration",
+        ).value
+        date_format = "%Y-%m-%d"
 
         try:
             datetime.strptime(date_string, date_format)
```
Comment (peaches7083) [inline coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py:280]: All this assertion proves is that the `date_string` is a date.  It doesn't prove that a leap year date (February 29th) was properly handled (-> March 31st) (?).

Comment (cdeanj2) [reply coldfront/plugins/qumulo/tests/management/test_billing_cycle_manager.py:280]: Good catch! I will fix that

File: coldfront/plugins/qumulo/management/commands/check_billing_cycles.py [hunk 8]
```diff
@@ -61,15 +62,18 @@ def calculate_prepaid_expiration(
...
+        last_day_of_month = False
         prepaid_months = int(prepaid_months)
-        prepaid_until = datetime(
-            prepaid_billing_start.year
-            + (prepaid_billing_start.month + prepaid_months - 1) // 12,
-            (prepaid_billing_start.month + prepaid_months - 1) % 12 + 1,
-            prepaid_billing_start.day,
-        )
+        prepaid_until = prepaid_billing_start + relativedelta(months=prepaid_months)
+        last_day_of_month = (prepaid_billing_start.day == calendar.monthrange(prepaid_billing_start.year, prepaid_billing_start.month)[1])
```
Comment (harterjwustl) [inline coldfront/plugins/qumulo/management/commands/check_billing_cycles.py:71]:
```suggestion
        is_last_day_of_month = (prepaid_billing_start.day == calendar.monthrange(prepaid_billing_start.year, prepaid_billing_start.month)[1])
```

Reason: Makes it clear variable is a boolean that answers a questions and not a number that stores information

**Review by gunnarwustl — APPROVED** [review]
